query/hash_reduction.cu

```diff
@@ -0,0 +1,395 @@
+
+    auto rawMapKeyIter = thrust::make_zip_iterator(
+        thrust::make_tuple(hashIter, thrust::counting_iterator<uint32_t>(0)));
+
+    auto hashIndexFusionFunc = []
+    __host__ __device__(
+        typename decltype(rawMapKeyIter)::value_type tuple) {
+      return (static_cast<int64_t>(thrust::get<0>(tuple)) << 32)
+          | (static_cast<int64_t>(thrust::get<1>(tuple)));
+    };
+
+    auto mapKeyIter = thrust::make_transform_iterator(rawMapKeyIter,
+                                                      hashIndexFusionFunc);
+
+    auto mapKeyValuePairIter = thrust::make_zip_iterator(
+        thrust::make_tuple(
+            mapKeyIter,
+            reinterpret_cast<value_type *>(inputValues)));
+
+    auto equality = []
+    __host__ __device__(key_type lhs, key_type rhs) {
+      return lhs >> 32 == rhs >> 32;
+    };
+
+    auto insertionFunc = [=]
+    __host__ __device__(
+        thrust::tuple<key_type, value_type> key_value_tuple) {
+      map->insert(tuple2pair(key_value_tuple), f, equality);
+    };
+
+    thrust::for_each_n(GET_EXECUTION_POLICY(cudaStream),
+                       mapKeyValuePairIter,
+                       length,
+                       insertionFunc);
+  }
+
+  template<typename map_type>
+  int output(map_type *map,
+             DimensionVector inputKeys,
+             DimensionVector outputKeys, uint8_t *outputValue,
+             cudaStream_t cudaStream) {
+    // calculate prefix sum of NumDimsPerDimWidth so that it will be easier
+    // to tell the valueBytes given a dim index. The prefix sum is inclusive.
+    uint8_t
+    dimWidthPrefixSum[NUM_DIM_WIDTH];
+    for (int i = 0; i < NUM_DIM_WIDTH; i++) {
+      dimWidthPrefixSum[i] = inputKeys.NumDimsPerDimWidth[i];
+      if (i > 0) {
+        dimWidthPrefixSum[i] += dimWidthPrefixSum[i - 1];
+      }
+    }
+
+    device_unique_ptr<uint32_t> globalWriteIndexDevice =
+        make_device_unique<uint32_t>(cudaStream);
+
+    ExtractGroupByResultFunctor<map_type> extractorFunc(
+        map, inputKeys.DimValues, inputKeys.VectorCapacity,
+        outputKeys.DimValues, dimWidthPrefixSum,
+        outputValue, globalWriteIndexDevice.get());
+
+    thrust::for_each_n(GET_EXECUTION_POLICY(cudaStream),
+                       thrust::counting_iterator<uint32_t>(0),
+                       length, extractorFunc);
+    // copy the reduced count back from GPU.
+    uint32_t
+    globalWriteIndexHost;
+    ares::asyncCopyDeviceToHost(
+        reinterpret_cast<void *>(&globalWriteIndexHost),
+        reinterpret_cast<void *>(globalWriteIndexDevice.get()),
+        sizeof(uint32_t), cudaStream);
+    return globalWriteIndexHost;
+  }
+
+  // execute reduces the dimension values with measures first and then extract
+  // aggregated result.
+  int execute(DimensionVector inputKeys, uint8_t *inputValues,
+              DimensionVector outputKeys, uint8_t *outputValues,
+              cudaStream_t cudaStream) {
+    auto equality = []
```

Comment:
yes,  but since it's lambda so we cannot store it as a member function. 
And since the equality check logic is simple so I think it's ok to duplicate the logic. 

This is actually a problem of this concurrent hash map, on L297 of the concurrent_map.cuh
```
typename hash_value_type = typename Hasher::result_type
```
it requires result type of the Hasher, however since our hasher is a lambda, it does not have a hash_value_type bound with it.

Anyway I think to some days we probably need to rewrite the hash map ourselves.